File: backend/connectors/reddit.py
"""Reddit API connector using asyncpraw."""

import logging
import re
from datetime import date

import asyncpraw

from backend.connectors.base import PlatformConnector, MetricResult

logger = logging.getLogger(__name__)


class RedditConnector(PlatformConnector):
    platform = "reddit"

    def __init__(
        self,
        client_id: str = "",
        client_secret: str = "",
        username: str = "",
        password: str = "",
    ):
        self.client_id = client_id
        self.client_secret = client_secret
        self.username = username
        self.password = password

    def _has_credentials(self) -> bool:
        return bool(self.client_id and self.client_secret)

    def _create_reddit(self) -> asyncpraw.Reddit:
        kwargs = {
            "client_id": self.client_id,
            "client_secret": self.client_secret,
            "user_agent": "TrikeriMarketingEngine/0.1",
        }
        if self.username and self.password:
            kwargs["username"] = self.username
            kwargs["password"] = self.password
        return asyncpraw.Reddit(**kwargs)

    async def validate_credentials(self) -> bool:
        """Test Reddit API credentials via asyncpraw."""
        if not self._has_credentials():
            return False
        try:
            reddit = self._create_reddit()
            # Attempt a lightweight API call
            user = await reddit.user.me()
            await reddit.close()
            return True
        except Exception as e:
            logger.warning("Reddit credential validation failed: %s", e)
            try:
                await reddit.close()
            except Exception:
                pass
            return False

    async def fetch_post_metrics(self, platform_post_id: str, **kwargs) -> MetricResult:
        """Fetch metrics for a Reddit submission.

        Mapping:
          likes = score (net upvotes)
          comments = num_comments
          shares = num_crossposts
          views = view_count (if available, usually None for non-OC)
        """
        if not self._has_credentials():
            raise ValueError(
                "Reddit API credentials not configured. "
                "Set TRIKERI_REDDIT_CLIENT_ID and TRIKERI_REDDIT_CLIENT_SECRET."
            )

        reddit = self._create_reddit()
        try:
            submission = await reddit.submission(id=platform_post_id)
            await submission.load()

            return MetricResult(
                views=getattr(submission, "view_count", 0) or 0,
                likes=submission.score,
                comments=submission.num_comments,
                shares=getattr(submission, "num_crossposts", 0) or 0,
                fetched_via="api",
                snapshot_date=date.today(),
                custom_metrics={
                    "upvote_ratio": submission.upvote_ratio,
                    "subreddit": str(submission.subreddit),
                    "is_original_content": submission.is_original_content,
                    "over_18": submission.over_18,
                },
            )
        finally:
            await reddit.close()

    async def resolve_post_id(self, url: str) -> str | None:
        """Extract Reddit post ID from URL.

        Handles: reddit.com/r/subreddit/comments/ABC123/...
        Returns: ABC123
        """
        match = re.search(r"/comments/([a-zA-Z0-9]+)", url)
        return match.group(1) if match else None

    def is_api_trackable(self) -> bool:
        return True

File: src-tauri/src/server/connectors/reddit.rs
use async_trait::async_trait;
use tokio::sync::RwLock;
use crate::server::connectors::{PlatformConnector, MetricResult};

pub struct RedditConnector {
    client: reqwest::Client,
    client_id: String,
    client_secret: String,
    username: String,
    password: String,
    token: RwLock<Option<TokenData>>,
}

struct TokenData {
    access_token: String,
    expires_at: chrono::DateTime<chrono::Utc>,
}

impl RedditConnector {
    pub fn new(
        client: reqwest::Client,
        client_id: String,
        client_secret: String,
        username: String,
        password: String,
    ) -> Self {
        Self {
            client, client_id, client_secret, username, password,
            token: RwLock::new(None),
        }
    }

    async fn get_token(&self) -> anyhow::Result<String> {
        // Check cached token
        {
            let guard = self.token.read().await;
            if let Some(ref t) = *guard {
                if t.expires_at > chrono::Utc::now() {
                    return Ok(t.access_token.clone());
                }
            }
        }

        // Fetch new token
        let resp = self.client
            .post("https://www.reddit.com/api/v1/access_token")
            .basic_auth(&self.client_id, Some(&self.client_secret))
            .header("User-Agent", "MeemMarketing/0.1")
            .form(&[
                ("grant_type", "password"),
                ("username", &self.username),
                ("password", &self.password),
            ])
            .send()
            .await?;

        let data: serde_json::Value = resp.json().await?;
        let access_token = data["access_token"].as_str()
            .ok_or_else(|| anyhow::anyhow!("No access_token in Reddit response"))?
            .to_string();
        let expires_in = data["expires_in"].as_i64().unwrap_or(3600);

        let mut guard = self.token.write().await;
        *guard = Some(TokenData {
            access_token: access_token.clone(),
            expires_at: chrono::Utc::now() + chrono::Duration::seconds(expires_in - 60),
        });

        Ok(access_token)
    }
}

#[async_trait]
impl PlatformConnector for RedditConnector {
    fn platform(&self) -> &str { "reddit" }

    async fn validate_credentials(&self) -> bool {
        match self.get_token().await {
            Ok(token) => {
                let resp = self.client
                    .get("https://oauth.reddit.com/api/v1/me")
                    .bearer_auth(&token)
                    .header("User-Agent", "MeemMarketing/0.1")
                    .send().await;
                matches!(resp, Ok(r) if r.status().is_success())
            }
            Err(_) => false,
        }
    }

    async fn fetch_post_metrics(&self, platform_post_id: &str) -> anyhow::Result<MetricResult> {
        let token = self.get_token().await?;
        let url = format!("https://oauth.reddit.com/api/info?id=t3_{}", platform_post_id);
        let resp = self.client
            .get(&url)
            .bearer_auth(&token)
            .header("User-Agent", "MeemMarketing/0.1")
            .send().await?;

        let data: serde_json::Value = resp.json().await?;
        let post = &data["data"]["children"][0]["data"];

        if post.is_null() {
            anyhow::bail!("Reddit post {} not found", platform_post_id);
        }

        Ok(MetricResult {
            views: post["view_count"].as_i64().unwrap_or(0),
            likes: post["score"].as_i64().unwrap_or(0),
            comments: post["num_comments"].as_i64().unwrap_or(0),
            shares: post["num_crossposts"].as_i64().unwrap_or(0),
            custom_metrics: Some(serde_json::json!({
                "upvote_ratio": post["upvote_ratio"],
                "subreddit": post["subreddit"],
                "is_original_content": post["is_original_content"],
                "over_18": post["over_18"],
            })),
            fetched_via: "api".into(),
            ..Default::default()
        })
    }

    fn resolve_post_id(&self, url: &str) -> Option<String> {
        // Extract from /comments/ABC123/
        let re_pattern = "/comments/([a-zA-Z0-9]+)";
        if let Some(start) = url.find("/comments/") {
            let rest = &url[start + 10..];
            let end = rest.find('/').unwrap_or(rest.len());
            let id = &rest[..end];
            if !id.is_empty() {
                return Some(id.to_string());
            }
        }
        let _ = re_pattern; // suppress unused warning
        None
    }

    fn is_api_trackable(&self) -> bool { true }
}
